graphql_compiler/compiler/emit_sql.py

```diff
@@ -205,81 +354,199 @@ def _set_group_by(self, group_by):
+                                 u'Cannot visit output vertices after end_fold has been '
+                                 u'called.'.format(self))
         if self._output_vertex_alias is not None:
-            raise AssertionError('Cannot visit multiple output vertices in one fold.')
+            raise AssertionError(u'Invalid state encountered for fold: {}'
+                                 u'Cannot visit multiple output vertices in one'
+                                 u' fold.'.format(self))
         self._output_vertex_alias = output_alias
         self._outputs = self._get_fold_outputs(fold_scope_location,
-                                               join_descriptor,
                                                all_folded_outputs)
 
     def visit_traversed_vertex(self, join_descriptor, from_table, to_table):
-        """Add a new join descriptor for every vertex traversed in the fold."""
-        self._join_info.append((join_descriptor, from_table, to_table))
+        """Add a new traversal descriptor for every vertex traversed in the fold."""
+        if self._ended:
+            raise AssertionError(u'Cannot visit traversed vertices after end_fold has been called.')
+        self._traversal_descriptors.append(SQLFoldTraversalDescriptor(join_descriptor,
+                                                                      from_table,
+                                                                      to_table))
 
     def end_fold(self, alias_generator, from_clause, outer_from_table):
         """Produce the final subquery and join it onto the rest of the query."""
+        if self._ended:
+            raise AssertionError(u'Invalid state encountered for fold: {}'
+                                 u'Cannot call end_fold more than once.'.format(self))
         if self._output_vertex_alias is None:
-            raise AssertionError('No output vertex visited.')
-        if len(self._join_info) == 0:
-            raise AssertionError('No traversed vertices visited.')
+            raise AssertionError(u'Invalid state encountered for fold: {}'
+                                 u'No output vertex visited.'.format(self))
+        if len(self._traversal_descriptors) == 0:
+            raise AssertionError(u'Invalid state encountered for fold: {}'
+                                 u'No traversed vertices visited.'.format(self))
 
         # for now we only handle folds containing one traversal (i.e. join)
-        if len(self.join_info) > 1:
-            raise NotImplementedError('Folds containing multiple traversals are not implemented.')
-        edge, from_alias, to_alias = self.join_info.pop()
+        if len(self._traversal_descriptors) > 1:
+            raise NotImplementedError(u'Not implemented error encountered for fold: {}'
+                                      u'Folds containing multiple traversals are not '
+                                      u'implemented.'.format(self))
```

Comment:
Do not start a `NotImplementedError`  with "Not implemented error".  

For `AssertionError`, I think that it is ok that you include as much information as you'd like. They indicate a bug and we need as much information as possible to debug. 

However, `NotImplementedError` are errors that are likely going to be seen multiple times by the users so please make sure that they make sense to someone with little to no understanding of how fold internally works.

Reply:
The same applies to your other `NotImplementedError`.